File: apps/api/src/agents/script-generator.agent.ts
type GenerateScriptInput = {
  baseUrl: string;
  flowName: string;
  steps: string[];
};

export async function generateScriptForFlow(input: GenerateScriptInput): Promise<string> {
  const safeName = input.flowName.replace(/[^a-zA-Z0-9 ]/g, '');

  if (/login/i.test(input.flowName)) {
    return `import { test, expect } from '@playwright/test';

test('${safeName}', async ({ page }) => {
  await page.goto('${input.baseUrl}');

  // Intentionally starts with a brittle selector to demonstrate auto-healing.
  await page.locator('#login-button').click();

  await page.getByLabel(/email/i).fill('demo@example.com');
  await page.getByLabel(/password/i).fill('password123');
  await page.getByRole('button', { name: /submit|login|sign in/i }).click();

  await expect(page.getByRole('heading', { name: /dashboard/i })).toBeVisible();
});
`;
  }

  return `import { test, expect } from '@playwright/test';

test('${safeName}', async ({ page }) => {
  await page.goto('${input.baseUrl}');
  await page.getByRole('textbox', { name: /search/i }).fill('browser automation');
  await page.keyboard.press('Enter');
  await expect(page.getByText(/results/i)).toBeVisible();
});
`;
}

File: apps/api-python/agents/script_generator_agent.py
from pathlib import Path
from langchain_core.prompts import ChatPromptTemplate
from core.config import settings
from core.llm import get_llm

def script_generator_node(state):
    llm = get_llm()
    prompt = ChatPromptTemplate.from_messages([
        ("system", "Generate Python Playwright pytest code. Prefer semantic locators. Return only valid Python."),
        ("human", "Base URL: {base_url}\nFlow: {flow}"),
    ])
    response = (prompt | llm).invoke({"base_url": state["base_url"], "flow": state["flow"]})

    script = f"""
from playwright.sync_api import sync_playwright, expect


def test_login_flow():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()
        page.goto("{state['base_url']}")

        # Brittle selector intentionally used for self-healing demo
        page.locator("#login-button").click()

        page.get_by_label("Email").fill("demo@example.com")
        page.get_by_label("Password").fill("password123")
        page.get_by_role("button", name="Submit").click()
        expect(page.get_by_role("heading", name="Dashboard")).to_be_visible()
        browser.close()
"""
    tests_dir = Path(settings.GENERATED_TESTS_DIR)
    tests_dir.mkdir(parents=True, exist_ok=True)
    script_path = tests_dir / "login_flow_test.py"
    script_path.write_text(script, encoding="utf-8")
    state["generated_script"] = script
    state["script_path"] = str(script_path)
    state["llm_generated_script_notes"] = response.content
    return state
